pkg/activator/breaker.go

```diff
@@ -0,0 +1,177 @@
+package activator
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	activatornetwork "github.com/knative/serving/pkg/activator/network"
+	"github.com/knative/serving/pkg/apis/networking"
+	netlisters "github.com/knative/serving/pkg/client/listers/networking/v1alpha1"
+	"github.com/knative/serving/pkg/network"
+	"github.com/knative/serving/pkg/network/prober"
+	"github.com/knative/serving/pkg/queue"
+	"go.uber.org/zap"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+)
+
+// Breaker is a wrapper around queue.Breaker which probes a revision before allowing breaker concurrency to rise from zero
+type Breaker struct {
+	*queue.Breaker
+	maxCC           int
+	logger          *zap.SugaredLogger
+	size            int
+	cc              int
+	replicaCount    int
+	isShutdown      bool
+	revId           RevisionID
+	mux             sync.Mutex
+	sizeChangedCond sync.Cond
+
+	sksLister     netlisters.ServerlessServiceLister
+	serviceLister corev1listers.ServiceLister
+	transport     http.RoundTripper
+}
+
+// NewBreaker creates a new Breaker
+func NewBreaker(
+	logger *zap.SugaredLogger,
+	transport http.RoundTripper,
+	breakerParams queue.BreakerParams,
+	replicaCount int,
+	sksLister netlisters.ServerlessServiceLister,
+	serviceLister corev1listers.ServiceLister,
+	revId RevisionID) *Breaker {
+	ret := &Breaker{
+		Breaker:      queue.NewBreaker(breakerParams),
+		maxCC:        breakerParams.MaxConcurrency,
+		logger:       logger,
+		size:         0,
+		cc:           0,
+		replicaCount: replicaCount,
+		isShutdown:   false,
+		revId:        revId,
+
+		sksLister:     sksLister,
+		serviceLister: serviceLister,
+		transport:     transport,
+	}
+	ret.sizeChangedCond.L = &ret.mux
+
+	// Run concurrency updating loop which polls endpoints
+	go func() {
+		ret.mux.Lock()
+		defer ret.mux.Unlock()
+
+		for {
+			if !ret.probeAndSetConcurrency() {
+				return
+			}
+
+			ret.sizeChangedCond.Wait()
+		}
+	}()
+
+	return ret
+}
+
+// Shutdown should be called on a breaker before it is forgotten to exit the probe goroutine
+func (b *Breaker) Shutdown() {
+	b.mux.Lock()
+	defer b.mux.Unlock()
+	b.isShutdown = true
+	// Break out of pollLoop wait
+	b.sizeChangedCond.Broadcast()
+}
+
+func (b *Breaker) probeAndSetConcurrency() bool {
+	for {
```

Comment:
So we're running an infinite loop in an infinite loop. I guess that is related to our use of `sync.Cond` here?